verifiers/internal/gcb/provenance.go

```diff
@@ -0,0 +1,308 @@
+	}
+	return d, nil
+}
+
+func (self *GCBProvenance) VerifyIntotoHeaders() error {
+	if err := self.isVerified(); err != nil {
+		return err
+	}
+
+	statement := self.verifiedIntotoStatementStruct
+	// https://in-toto.io/Statement/v0.1
+	if statement.StatementHeader.Type != intoto.StatementInTotoV01 {
+		return fmt.Errorf("%w: expected statement header type '%s', got '%s'",
+			serrors.ErrorInvalidDssePayload, intoto.StatementInTotoV01, statement.StatementHeader.Type)
+	}
+
+	// https://slsa.dev/provenance/v0.1
+	if statement.StatementHeader.PredicateType != slsa01.PredicateSLSAProvenance {
+		return fmt.Errorf("%w: expected statement predicate type '%s', got '%s'",
+			serrors.ErrorInvalidDssePayload, slsa01.PredicateSLSAProvenance, statement.StatementHeader.PredicateType)
+	}
+
+	return nil
+}
+
+func isValidBuilderID(id string) error {
+	for _, b := range GCBBuilderIDs {
+		if id == b {
+			return nil
+		}
+	}
+	return serrors.ErrorMismatchBuilderID
+}
+
+func (self *GCBProvenance) VerifyBuilderID(builderOpts *options.BuilderOpts) (string, error) {
+	if err := self.isVerified(); err != nil {
+		return "", err
+	}
+
+	statement := self.verifiedIntotoStatementStruct
+	predicateBuilderID := statement.Predicate.Builder.ID
+
+	// Sanity check the builderID.
+	if err := isValidBuilderID(predicateBuilderID); err != nil {
+		return "", err
+	}
+
+	// Validate with user-provided value.
+	if builderOpts != nil && builderOpts.ExpectedID != nil {
+		if *builderOpts.ExpectedID != predicateBuilderID {
+			return "", fmt.Errorf("%w: expected '%s', got '%s'", serrors.ErrorMismatchBuilderID,
+				*builderOpts.ExpectedID, predicateBuilderID)
+		}
+	}
+
+	// Valiate that the recipe type is consistent.
+	if predicateBuilderID != statement.Predicate.Recipe.Type {
+		return "", fmt.Errorf("%w: expected '%s', got '%s'", serrors.ErrorMismatchBuilderID,
+			predicateBuilderID, statement.Predicate.Recipe.Type)
+	}
+
+	// Validate the recipe argument type.
+	expectedType := "type.googleapis.com/google.devtools.cloudbuild.v1.Build"
+	args, ok := statement.Predicate.Recipe.Arguments.(map[string]interface{})
+	if !ok {
+		return "", fmt.Errorf("%w: recipe arguments is not a map", serrors.ErrorInvalidDssePayload)
+	}
+	ts, err := getAsString(args, "@type")
+	if err != nil {
+		return "", err
+	}
+
+	if ts != expectedType {
+		return "", fmt.Errorf("%w: expected '%s', got '%s'", serrors.ErrorMismatchBuilderID,
+			expectedType, ts)
+	}
+
+	return predicateBuilderID, nil
+}
+
+func getAsString(m map[string]interface{}, key string) (string, error) {
+	t, ok := m["@type"]
+	if !ok {
+		return "", fmt.Errorf("%w: '%s' field is absent", serrors.ErrorInvalidDssePayload, key)
+	}
+	ts, ok := t.(string)
+	if !ok {
+		return "", fmt.Errorf("%w: '%s' is not a string", serrors.ErrorInvalidDssePayload, key)
+	}
+	return ts, nil
+}
+
+func (self *GCBProvenance) VerifySubjectDigest(expectedHash string) error {
```

Comment:
The wrapped gcpProvenance headers are unverified, should we leave them that way, or verify against these fields against the imagesummary as well?

Reply:
I would like to verify them. Either in this PR or a follow-up PR. Thanks for the reminder.
I've been hesitant to do it till now because I don't know what the integration with gcloud will look like. As soon as I get more insight into this I'll add it.
Created https://github.com/slsa-framework/slsa-verifier/issues/221 for tracking

Reply:
Added verification for the summary and the metadata. I will do the humand-readable part later. The reason is that it does not follow the intoto format (`slsaProvenance` instead of `predicate`), so I'd like GCB team to fix that

Reply:
Added to https://github.com/slsa-framework/slsa-verifier/issues/221.